vaadin-upload-flow-parent/vaadin-upload-flow/src/main/java/com/vaadin/flow/component/upload/Upload.java

```diff
@@ -727,14 +731,7 @@ public void setReceiver(Receiver receiver) {
      *            upload handler to use for file receptions
      */
     public void setUploadHandler(UploadHandler handler) {
-        UploadHandler newUploadHandler = event -> {
-            try {
-                startUpload();
-                handler.handleUploadRequest(event);
-            } finally {
-                endUpload();
-            }
-        };
+        UploadHandler newUploadHandler = new UploadHandlerWrapper(handler);
```

Comment:
The drawbacks of this are:
- once any new method is added into `UploadHandler`, we should modify this code here accordingly.
- delegate for `handleRequest` is skipped, because the default impl uses `this` when it invokes `TransferUtils` and `this` means the original handler.

Any better solutions are welcomed.

Reply:
The only idea for nicer usage would be if the handler would fire events that you could add a listener for.
e.g. `handler.addEventListener( eventType -> if (eventType.equals(UploadEvent.START) { startUpload();} ... );`
Then anyone requiring before after etc could just listen for those instead of doing multiple wrappers hiding the actual implementation of the handler.

Reply:
You are now talking about `TransferProgressListener::onStart` and `TransferProgressListener::onComplete` basically. The obstacle for this approach is that we decided at some point to not expose `addTransferProgressListener` in handlers and, moreover, the `UploadHandler` functional interface has no API/implementation for events/listeners.

Reply:
No I meant specifically that the UploadHandler would support events so not tied to the transfer in any way.

Reply:
But any which way it is done the need would be to be able to have a way to do something immediately when `TransferUtil.handleUpload` is called and just before it exits. or `UploadHandler.handleRequest` called/exits.